ecosystem/indexer-grpc/transaction-filter/src/traits.rs

```diff
@@ -0,0 +1,202 @@
+            Some(filter) => filter.is_valid(),
+            None => Ok(()),
+        }
+    }
+
+    #[inline]
+    fn is_allowed(&self, item: &T) -> bool {
+        match self {
+            Some(filter) => filter.is_allowed(item),
+            None => true,
+        }
+    }
+
+    #[inline]
+    fn is_allowed_opt(&self, item: &Option<T>) -> bool {
+        match self {
+            Some(filter) => filter.is_allowed_opt(item),
+            None => true,
+        }
+    }
+}
+
+impl Filterable<String> for Option<String> {
+    #[inline]
+    fn validate_state(&self) -> Result<(), FilterError> {
+        Ok(())
+    }
+
+    #[inline]
+    fn is_allowed(&self, item: &String) -> bool {
+        match self {
+            Some(filter) => filter == item,
+            None => true,
+        }
+    }
+}
+
+impl Filterable<i32> for Option<i32> {
+    #[inline]
+    fn validate_state(&self) -> Result<(), FilterError> {
+        Ok(())
+    }
+
+    #[inline]
+    fn is_allowed(&self, item: &i32) -> bool {
+        match self {
+            Some(filter) => filter == item,
+            None => true,
+        }
+    }
+}
+
+impl Filterable<bool> for Option<bool> {
+    #[inline]
+    fn validate_state(&self) -> Result<(), FilterError> {
+        Ok(())
+    }
+
+    #[inline]
+    fn is_allowed(&self, item: &bool) -> bool {
+        match self {
+            Some(filter) => filter == item,
+            None => true,
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use anyhow::anyhow;
+
+    #[derive(Debug, Serialize, PartialEq)]
+    pub struct InnerStruct {
+        pub a: Option<String>,
+    }
+
+    impl Filterable<InnerStruct> for InnerStruct {
+        fn validate_state(&self) -> Result<(), FilterError> {
+            Err(anyhow!("This is a test error!").into())
+        }
+
+        fn is_allowed(&self, _item: &InnerStruct) -> bool {
+            true
+        }
+    }
+
+    #[derive(Debug, PartialEq, Serialize)]
+    pub struct OuterStruct {
+        pub inner: Option<InnerStruct>,
+    }
+
+    impl Filterable<InnerStruct> for OuterStruct {
+        fn validate_state(&self) -> Result<(), FilterError> {
+            self.inner.is_valid()?;
+            Ok(())
+        }
+
+        fn is_allowed(&self, item: &InnerStruct) -> bool {
+            self.inner.is_allowed(item)
+        }
+    }
+
+    #[test]
+    fn test_error_prop() {
+        let inner = InnerStruct {
+            a: Some("test".to_string()),
+        };
+        let outer = OuterStruct { inner: Some(inner) };
+
+        let res = outer.is_valid();
+        assert!(res.is_err());
+        let error = res.unwrap_err();
+        assert_eq!(error.to_string(), "Filter Error: This is a test error!\nTrace Path:\naptos_transaction_filter::traits::test::InnerStruct:   {\"a\":\"test\"}\ncore::option::Option<aptos_transaction_filter::traits::test::InnerStruct>:   {\"a\":\"test\"}\naptos_transaction_filter::traits::test::OuterStruct:   {\"inner\":{\"a\":\"test\"}}");
```

Comment:
Nit, to make this more readable can we use a raw string or indoc or something for the assertion?